### PR TITLE
Adds limb stuns status effects

### PR DESCRIPTION
**Pull request by colossusqw**
[LABEL][feature][balance][chemistry]

## About the PR 
This PR adds limb stuns, which functionally make the stunned limb act like it isn't there at all. Significantly less punishing than a straight up KO or general stun, but dinamic in effects and causes.

Hackmd link: https://hackmd.io/@colossusqw/HJ1Wnurb0

At the moment the effect is added to:
- Cyanide, which has a progressive set of random limb stun chances.
- Pancuronium, which progressively shuts down each of the player's limbs before paralysis.
- Explosions, which now have a chance of numbing a limb instead of blowing it off, as a middle ground.
- Glass shards, which temporarily stun a leg instead of full on stunning.
- Touching things that are electrified, which stuns the hand you used to touch it with.
image

## Why's this needed? 
Random stuns tend to be a bit RNG-y and frustrating, considering they have very little counterplay, this aims to address it by reducing a players functioning amount of limbs temporarily without straight up taking them out of the fight.

As for cyanide, the random stuns don't exactly make sense, so temporarily losing control of limbs is also more akin to cyanide's exceedingly strong muscle spasms. Pancuronium is made slightly less instant, but still pretty debilitating in how it's effects ramp up to full paralysis.

For explosions, it feels nice to have a secondary, less all or nothing effect to them, which feels like a less permanent and punishing version of losing an entire limb.

```changelog
(u)Colossusqw
(*)Added specific limb stuns, see minor changes for more information.
(+) Cyanide now has it's random stuns replaced by limb stuns, and Pancuronium has been reworked to progressively shut off limbs before paralysis.
(+) Explosions now have a chance to numb a limb as a middle ground between limb severing and no effects. Glass shards now stun your legs instead of you.
```

### CODE REVIEW

**Comment by deathrobotpunch** [comment]
could you add limb stuns to ricin’s later stages to add more *pizaz* to the whole organs shutting down thing

**Comment by colossusqw** [comment]
> could you add limb stuns to ricin’s later stages to add more _pizaz_ to the whole organs shutting down thing

That does make sense, yeah. At the moment it's just cyanide but i do want to incorporate it in place of different stuns that feel a bit cheap/rng-y around, ricin might be a good one to look at.

**Comment by deathrobotpunch** [comment]
a list of things this could also be added to are: the early stages of medusa, space madness, neurodepressant, saxitoxin, RNG on brain damage and withdrawal (stimulants and addictions)

**Comment by colossusqw** [comment]
!merge_upstream

File: code/datums/hud/human.dm [hunk 1]
```diff
@@ -780,12 +780,18 @@
 		if (master.limbs && !master.limbs.l_arm)
 			lhand.icon_state = "handl[master.hand]d"
 		else
-			lhand.icon_state = "handl[master.hand]"
+			if (master.limbs && master.hasStatus("numb_l_arm"))
+				lhand.icon_state = "handl[master.hand]s"
```
Comment (pgmzeta) [inline code/datums/hud/human.dm:784]: why can't we use overlays for the big X?

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 2]
```diff
@@ -333,6 +333,33 @@ datum
 					holder.remove_reagent(reagent_id, (amount * holder.reagent_list[reagent_id].flushing_multiplier))
 			return
 
+		//proc for stunning multiple limbs without having them overlap or stunning inexistent limbs
+		proc/numb_limb(var/mob/M, var/numbness_time, var/numb_amount = 1, var/random_order = TRUE, var/list/numb_limbs = list("numb_l_arm", "numb_l_leg", "numb_r_arm", "numb_r_leg"))
```
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:337]: numb in function name, can slim down the signature
```suggestion
		proc/numb_limb(mob/M, duration, count=1, random = TRUE, list/target_limbs= list("l_arm", "l_leg", "r_arm", "r_leg"),)
```

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 3]
```diff
@@ -333,6 +333,33 @@ datum
 					holder.remove_reagent(reagent_id, (amount * holder.reagent_list[reagent_id].flushing_multiplier))
 			return
 
+		//proc for stunning multiple limbs without having them overlap or stunning inexistent limbs
```
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:336]:
```suggestion
		///Numb one or more limbs without stacking status effects or targeting limbs that don't exist
```

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 4]
```diff
@@ -333,6 +333,33 @@ datum
...
+		proc/numb_limb(var/mob/M, var/numbness_time, var/numb_amount = 1, var/random_order = TRUE, var/list/numb_limbs = list("numb_l_arm", "numb_l_leg", "numb_r_arm", "numb_r_leg"))
+			if(!ishuman(M))
+				return
+			var/mob/living/carbon/human/H = M
+			if(random_order)
```
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:341]:
```suggestion	
			if(!H.limbs)
				return // no limb datum
			var/list/limbs_to_numb = list()
			for(var/target in target_limbs)
				var/obj/item/parts/limb = H.limbs.vars[target]
				if(istype(limb))
					limbs_to_numb += limb
			if(!length(limbs_to_numb))
				return // no limbs
			if(random_order)
```

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 5]
```diff
@@ -333,6 +333,33 @@ datum
...
+				if (i == "numb_l_leg" && H.limbs?.l_leg)
+					M.setStatusMin(i, numbness_time)
+					continue
+				if (i == "numb_r_leg" && H.limbs?.r_leg)
+					M.setStatusMin(i, numbness_time)
+					continue
+				numb_amount +=1 //does not register a stunned limb if none is stunned, allowing it to find another
+								//or stopping as soon as it runs out of possible limbs to stun
+			return
+
```
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:362]: pre-filter streamlines for loop
```suggestion
			for (var/i in 1 to min(count, length(limbs_to_numb)))
				M.setStatusMin("numb_[limbs_to_numb[i]]", duration) // i think this works but check
```

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 6]
```diff
@@ -333,6 +333,33 @@ datum
 					holder.remove_reagent(reagent_id, (amount * holder.reagent_list[reagent_id].flushing_multiplier))
 			return
 
+		//proc for stunning multiple limbs without having them overlap or stunning inexistent limbs
+		proc/numb_limb(var/mob/M, var/numbness_time, var/numb_amount = 1, var/random_order = TRUE, var/list/numb_limbs = list("numb_l_arm", "numb_l_leg", "numb_r_arm", "numb_r_leg"))
+			if(!ishuman(M))
+				return
+			var/mob/living/carbon/human/H = M
```
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:340]: Pre-filter the limb list

```suggestion
			var/mob/living/carbon/human/H = M
			if(!H.limbs)
				return // no limb datum
			var/list/limbs_to_numb = list()
			for(var/target in target_limbs)
				var/obj/item/parts/limb = H.limbs.vars[target]
				if(istype(limb))
					limbs_to_numb += limb
			if(!length(limbs_to_numb))
				return // no limbs
```

**Review by pgmzeta — CHANGES_REQUESTED** [review]
Removing limbs should clear the associated status effect: https://github.com/goonstation/goonstation/blob/94f58ed5fddcbd46d516ce561f4b4c0f46000109/code/obj/item/mob_parts.dm#L153

File: code/modules/chemistry/Chemistry-Reagents.dm (hunk 5 above)
Comment (pgmzeta) [inline code/modules/chemistry/Chemistry-Reagents.dm:362]: cleaned up and tested suggestion

```suggestion
        ///Numb one or more limbs without stacking status effects or targeting limbs that don't exist
		proc/numb_limb(mob/M, duration, count=1, random=TRUE, list/target_limbs=list("l_arm", "l_leg", "r_arm", "r_leg"),)
			if(!ishuman(M))
				return
			var/mob/living/carbon/human/H = M
			if(!H.limbs)
				return // no limb datum
			var/list/obj/item/parts/limbs_to_numb = list()
			for(var/target in target_limbs)
				var/obj/item/parts/limb = H.limbs.vars[target]
				if(istype(limb))
					limbs_to_numb += limb
			if(!length(limbs_to_numb))
				return // no limbs
			if(random)
				shuffle_list(limbs_to_numb)
			for (var/i in 1 to min(count, length(limbs_to_numb)))
				M.setStatusMin("numb_[limbs_to_numb[i].slot]", duration)
```

**Comment by flappybatpal** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge-upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
That commit name is wrong, didn't add the effects to any other chems yet, still just the limb stuns

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 7]
```diff
@@ -333,6 +333,24 @@ datum
...
+			for(var/target in target_limbs)
+				var/obj/item/parts/limb = H.limbs.vars[target]
+				if(istype(limb))
+					limbs_to_numb += limb
+			if(!length(limbs_to_numb))
+				return // no limbs
+			if(random)
+				shuffle_list(limbs_to_numb)
+			for (var/i in 1 to min(count, length(limbs_to_numb)))
+				M.setStatusMin("numb_[limbs_to_numb[i].slot]", duration)
```
Comment (Tarmunora) [inline code/modules/chemistry/Chemistry-Reagents.dm:353]: why is this a proc on reagent

Comment (colossusqw) [reply code/modules/chemistry/Chemistry-Reagents.dm:353]: It allows for customizability and ease of application.
A good example would be a set choice of non-random limb stuns, stunning 2 different limbs at once or setting only arms or legs for stunning, without having to apply all the logic in each possible instance in the future.

Comment (Tarmunora) [reply code/modules/chemistry/Chemistry-Reagents.dm:353]: But why is this on reagent and not on human?

Comment (colossusqw) [reply code/modules/chemistry/Chemistry-Reagents.dm:353]: Oh, wow i don't think i realized that. Frankly it's been so long that i made this proc that i don't quite recall my reason for putting it on reagent. I can definitely change it though.

Comment (colossusqw) [reply code/modules/chemistry/Chemistry-Reagents.dm]: Moved to human.dm

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
T-T stale once more

**Review by ZeWaka — COMMENTED** [review]
i'm not sure about a lot of the code logic in this pr

**Comment by colossusqw** [comment]
> i'm not sure about a lot of the code logic in this pr

Any particular things you'd like to see changed? It didn't seem to cause any bugs or weird behavior in practice on the if anything isn't up to standard, I can try changing it.

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Comment by colossusqw** [comment]
!merge_upstream

**Pull request closed without merging**